src/Utilities/System/Prefetch.hpp

```diff
@@ -0,0 +1,38 @@
+// Distributed under the MIT License.
+// See LICENSE.txt for details.
+
+#pragma once
+
+#include <iosfwd>
+
+namespace sys {
+/// The cache location to prefetch data to.
+enum class PrefetchTo : int {
+  /// Prefetch into L1 data cache
+  L1Cache = 3,
+  /// Prefetch into L2 data cache
+  L2Cache = 2,
+  /// Prefetch into L3 data cache
+  L3Cache = 1,
+  /// Non-temporal is an element that is unlikely to be re-used. E.g., read
+  /// once, written but never read.
+  NonTemporal = 0,
+  /// Prefetch to L1 data cache for writing
+  WriteL1Cache = 7,
+  /// Prefetch to L2 data cache for writing
+  WriteL2Cache = 6
+};
+
+/// \brief Prefetch data into a specific level of data cache.
+template <PrefetchTo CacheLocation>
+#if defined(__GNUC__)
+__attribute__((always_inline)) inline
+#endif
+void prefetch(const void* address_to_prefetch) {
+  __builtin_prefetch(address_to_prefetch,
+                     (static_cast<int>(CacheLocation) >> 2) & 1,
+                     static_cast<int>(CacheLocation) & 0x3);
```

Comment:
I'm not sure what you want me to say here. This is manipulated to it givens generally the write numbers for compilers. I could say something like "Manipulate enum value to pass to compiler" but that's sort of redundant with the code. What are you looking for?

Reply:
I think what @nilsvu is looking for (and what I was wondering), is why are the enum values what they are? Why is L1Cache 3, why is L3Cache 1? A comment on that would be helpful. And then maybe also what the `>> 2` and `& 0x3` actually do? I know they are well defined operations, but they aren't common so a comment explaining what the operation returns would be helpful.

Reply:
I still don't understand, sorry. The values are chosen so that the corresponding bit manipulations are the correct values passed to `__builtin_prefetch` to load something into the correct level of cache. I don't think it's good to re-document `__builtin_prefetch` since GCC (and presumably Clang) document it. I could use something like a switch or if-constexpr but that would, in the case of switch, be bad for performance since that's a runtime lookup, and if-constexpr would create a very long function.

I don't think we should document standard operators like bitwise and right shift in the code.

The point of the wrapper is to hide the details of `__builtin_prefetch`, so I'm not sure why we are focusing on the details of `__builtin_prefetch`. It's not the user-interface. The enum is supposed to be a helpful representation of bit flags. Would it be useful to write the values in the enum in binary format?

Reply:
What about:
```
// The enum values are bit flags where the lowest two bits (right-most) control
// the cache level and the 3rd bit controls whether it is a write-only or read-write operation.
```

Reply:
Sounds good 👍 That's enough for me

Reply:
Yeah that works for me as well